### DISCUSSION TITLE
Thinking Through Quest Groups

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3209] [Ideas]
### Goals

**I would like your help refining the idea of quest groups.** It is a pattern that could be used in/alongside many different proposals, including presets. It has been floating around in my mind for a while, but I haven't articulated it clearly until now (I hope :smile:).

### Explanation

I still don't have an airtight definition, so I will explain by example.

In the context of sorting quests, groups might be something like: instead of showing separate entries for parking type, access, and fee which could be re-ordered individually, there is just one entry for "parking". To avoid getting hung up on details: Perhaps it can be expanded to still enable/disable individual quests. Perhaps groups would be user-customizable. The point is that they allow you to change (re-order, or enable/disable) multiple quests at once, not just one.

- https://github.com/streetcomplete/StreetComplete/issues/2565 are groups for disabling quests. "Quests that require you to go inside" were proposed as a non-customizable default group, and likely the user would be able to add custom groups.
- **[Favorites/low priority](https://github.com/streetcomplete/StreetComplete/issues/3034#issuecomment-879481130)** are groups for enabling/disabling *and* sorting. The user can choose which quests are in the groups, but not add custom groups. (note: the comment is hidden, you'll have to expand it)
- My original suggestion for **[quest priority selection on the intro screen](https://github.com/streetcomplete/StreetComplete/issues/1654#issuecomment-742040745) (the left image)** uses pre-defined groups used to both re-order or disable quests.
    - Much of https://github.com/streetcomplete/StreetComplete/discussions/2457 is about creating groups for use cases.

So, yeah— all of my previous suggestions were quest groups… which is how I noticed the concept to begin with. I currently suspect that I'll be happy with any form of quest groups and none of the other details matter, so **I'm happy to close all those previous threads**; this discussion supersedes anything I said in them.

Note: I don't count quest presets as an implementation of quest groups (even if they *technically* are). The reason is: they don't divide the quests. Every quest is in every group and there's no way to combine them (e.g. "show all the quests active in either of these presets"). However, they are *compatible* with groups: the order (or enabled/disabled status) of quests within a preset could be set using groups.

### Why I think it's a good idea

Aka, all the things you might disagree with. 

Note: I couldn't find a word to concisely represent the concept of "any feature that helps users answer the quests they're interested in— quest presets, ordering, display density, showing all quests for one point, etc." So I made up a word to refer to it concisely: QF (from "Quest customization Feature", but I'm using it to describe the concept, not a 1:1 substitution with that phrase).

#### Base assumptions

1. People have widely **different interests**; quests which are important to [some people](https://github.com/streetcomplete/StreetComplete/issues/3132#issuecomment-894827733) would be considered spam by others.
2. Showing too many unimportant quests will cause people to stop using StreetComplete.
3. Better QFs are **widely desired** (based on how often they are requested; see https://github.com/streetcomplete/StreetComplete/issues/1654#issuecomment-742077820, https://github.com/streetcomplete/StreetComplete/discussions/2789#discussioncomment-646707, https://github.com/streetcomplete/StreetComplete/issues/1914)
4. If QFs are too much effort to use (for whatever reason), people won't use them.

#### Conclusions

5. Better QF allows us to add more **diverse quests** without driving away users. This could make a **[huge improvment](https://github.com/streetcomplete/StreetComplete/discussions/2671#discussioncomment-783347)** in amount of data SC can gather (probably a gradual increase, rather than a huge spike like the offline data download).
6. …but only if the QFs are **simple enough** that most people will use them. If they are only used by the 5-10 people commenting here, it would be better to put them in a different [version of SC](https://github.com/streetcomplete/StreetComplete/discussions/3003).
7. As we continue adding more quests, the time required to manage them individually will **increase non-linearly**; on top of having more quests, each one will take longer to locate and move to the right priority.
8. The only way I see to satisfy the tradeoff between adding more quests and keeping QFs usable is the ability making QFs that are less fine-grained. In other words, quest groups.

### Misc

- Most of the use cases and specific proposals I've thought of are things I've already proposed.
- The [updated quest order](https://github.com/streetcomplete/StreetComplete/blob/f46cf9558685f2e58fbe4fae96bc875fd3cd221e/app/src/main/java/de/westnordost/streetcomplete/quests/QuestModule.kt) declarations are already separated into some groups.

**Comment by smichel17** [comment]
I wrote the first draft of this post a week ago, and have been rewriting (to try and avoid going in the direction of *yet another quest presets discussion*) and removing unnecessary parts since then. It's still not perfect (missing some ideas and a little longer than I'd like), but at some point I need to say "good enough" :P

**Comment by smichel17** [comment]
The tensions I have around quest groups are conflicts with other features.

For example, I think it would be a good idea if "prerequisite" quests (ex: building type is a prereq for house number) would be upgraded in priority to the highest quest that depends on them (building type quests *for buildings with no house number* should have the same priority as the house number quest, if it is higher). But this isn't needed if we use groups for prioritizing and make a group for building-related quests, since they wouldn't be far enough apart in the priority order for it to matter.

On one hand, this is good— groups are a simpler implementation to accomplish the same thing. But it feels like "perfect is the enemy of good" in reverse; implementing the "good enough" solution means the feature that requires no user interaction doesn't get built.

**Comment by mnalis** [comment]
So, something like https://github.com/streetcomplete/StreetComplete/issues/2308 ? Grouped quests **do** sound useful, but their implementation and user actions needed to make them useful is IMHO likely to be too complex (sorry to sound like a downer again :smile:)

Few thoughts that immediately come to mind:

- **UI needed** for creating and grouping quests to custom groups  seems **likely to be complex** (one needs to **at least** be able: to activate groups, deactivate groups, move entries between groups, create new groups, remove groups. Also might be useful to be able to collapse/expand groups, and rename existing groups). Moving quests between groups seems clunky - as can be seen as currently implemented in _"drag&drop"_ method for quest priorities - especially when you want to move it across multiple screens of quests (eg. from bottom to top). Having to do it again for groups is multiplying that pain significantly. _"cut&paste"_ method might be other option, but it has its own problems
- **how would quest priority sorting be implemented?** If the quests are shown inside groups, then that would not be able to show (much less allow changing) quest priorities; so one would have to have `group view` for managing content of groups, and `global view` to manage quest priorities. We need (at least I do) **per-quest priorities**, it does not help me if I can just (de-)prioritize whole `building` group for example (I absolutely want to prioritize `address`, but de-prioritize `roof shape` individually, for example)
- how would **quests that belong to multiple categories** be handled? (eg. `bicycle parking` might be in the `parking` group, but also in the `cycling` group, etc. By increased complexity, eg. having `duplicate quest` option as well as `remove this duplicate` option? Or it would be disallowed, thus severely reducing the usefulness of the groups - eg. no `inside` group as stuff is already in `food` group? Or would each group have full complement of all quests, and user would have "throw out" all quests unwanted in that group - and what if they throw out too much? Will there be undo?)
- the complexity of the `groups` would **add** to one already implemented for quests presets (https://github.com/streetcomplete/StreetComplete/tree/quest-profiles). That is something to keep in mind; as even setting up just 3 presets is no easy task. If there were dozen groups to manage too, it would be even harder.
- how would **new quests** (on SC upgrade) be handled? Would they automagically drop in some group (and which one, if so? And would the user be notified about that - and if so, how? Or would it happen silently so user might never see the new quests unless they are on alert and revisit/update all groups after each upgrade?)

At this point, I'm generally of the opinion to wait for next major release with `quest presets` implemented, and see how that works and how well it handles real-life use cases (for some time), and then eventually look what more (if anything) might be changed so usability is improved, while making sure the complexity stays reasonable.

**Reply by smichel17** [reply]
> user actions needed to make them useful is IMHO likely to be too complex (sorry to sound like a downer again :smile:)

No problem. I suspected the same thing, given how much I have struggled to come up with an implementation. A reply like this is exactly what I was hoping for.

I also agree that a **wait-and-see** approach makes the most sense at this time. So, this discussion is not very urgent. This is part of the reason why I have taken 10 days to respond (and I'll respond to the other parts gradually over time).

**Reply by smichel17** [reply]
> the complexity of the `groups` would **add** to one already implemented for quests presets

I agree that having a fully customizable implementation of groups *in addition* to quest presets would be *too complex.* So instead, here I'll talk about ideas to **replace** quest presets. Of course, that makes this a **theoretical** discussion since removing quest presets is unlikely to happen. But perhaps it will still lead somewhere interesting.

> one needs to at least be able: to activate groups, deactivate groups, move entries between groups, create new groups, remove groups. Also might be useful to be able to collapse/expand groups, and rename existing groups

Here is the simplest UI I can think of:

1. The **Manage Groups** screen. It can be opened from the **main menu** on the map screen. It looks like the current quest list. Each group has:

    - An **icon** and a name.
    - A handle on the left allowing you to drag and **reorder** it.
    - A checkbox on the right to **toggle** (enable/disable) that group.

   I'm not sure the best mechanism for deleting groups. Available options are long-pressing, swiping, or a menu item (`Remove groups`) that lets you tap groups to delete them. (edit: same for renaming)

   There would be two special groups, which cannot be deleted or renamed:

    - `No group` — This is where new quests are added and where most quests are by default.
    - `Disabled` — also cannot be re-ordered (it is always at the bottom) or enabled (it is always disabled, as the name suggests). Technically it doesn't need to exist, but I don't want people to accidentally remove their ability to disable quests from the quest list.

   One of the menu items would be a shortcut to open the quest list.

2. The **quest list** screen:

    - Quests may no longer be re-ordered (handle removed).
    - Quests may no longer be disabled (checkbox removed).
        - Replaced with the icon of the quest group it belongs to.
    - Tapping a quest opens a menu to change the quest group (or maybe cycles through groups directly, if there are only a couple)

So, there you have it. Each quest can only belong to one group. Instead of switching presets, you'd open up the quest groups screen and re-order or enable/disable groups. More rarely, you'd open the quest list and change which group each quest belongs to.

It's a bit more effort to change the order than merely selecting a preset, and a bit less flexible (unless you want to create an inordinate amount of groups, which would defeat the purpose). But I think it should be easier to set up, since you only have to assign quests to groups once, instead of once per preset.

I suppose technically presets could be layered on top, as a way to quickly re-order & toggle groups, but I don't think that would be a good trade-off in complexity/functionality.

**Comment by mnalis** [comment]
(There were also Layers idea https://github.com/streetcomplete/StreetComplete/issues/2461 which might be related, depending if it would support multiple related quests per layer / multiple active layers)

**Comment by mnalis** [comment]
> 2. Showing too many unimportant quests will cause people to stop using StreetComplete.

Agree. Less important should be disabled by default, as https://github.com/streetcomplete/StreetComplete/tree/quest-order already does it for some I think? Then only users who are bored (so likely do not mind extra complexity) would go into settings and enable extra lower-value quests. 

> 7. As we continue adding more quests, the time required to manage them individually will **increase non-linearly**; on top of having more quests, each one will take longer to locate and move to the right priority.

Technically, it will still **increase linearly**, so O(n) complexity. 
Eg. if new version had 500 quests instead of 100, it would take you 5 times as long to find a quest manually browsing the list, and then 5 times as long to move it where you want it. (But `y=a*x+b` is still a linear growth, even in `a=2` instead of `a=1`). (Actually much less if one uses `search quest` functionality to just enable/disable quest)

But do note that if the groups are not hardcoded but user-configurable (and I for one **only see utility if groups would be user configurable**, otherwise they would effectively totally negate usefulness of `user presets`!), there is still going to be that increased complexity on each upgrade (perhaps even more than without groups, depending on number of groups vs. number of presets).

> 8. The only way I see to satisfy the tradeoff between adding more quests and keeping QFs usable is the ability making QFs that are less fine-grained. In other words, quest groups.

Well, maybe it might be one way to do it, but see above - it might even increase complexity instead of reducing it. 

There might be other ways to make easier to detect and thus modify new quests (perhaps color new quests differently after upgrade?) which is the _biggest issue I see with growing number of quests for existing users_ (completely new users will IMHO have to either use the defaults - most of them I guess; or browse the full quest list anyway to see what else is there and enable/disable what quests they want exactly)

I agree with rest of the Base assumptions (1,3,4) and Conclusions (5,6) without extra comments.

**Reply by smichel17** [reply]
> Technically, it will still increase linearly, so O(n) complexity.

I had a comment here but lost it due to accidentally closing the browser (GitHub does a great job of preserving top level comments, but apparently not replies on discussions). Anyway, the gist was:

- I think you're talking about linear to reorder a single quest; I was saying it's nonlinear to reorder the entire list.
- With simplifying assumptions, like the default order being pretty good, it's linear or less.
- I should probably not have mentioned growth rate since it's just nitpicking; in practice the effort is *already* too much, so I could have made my point with simply "it will get (some amount) worse".

**Comment by smichel17** [comment]
> I for one **only see utility if groups would be user configurable**, otherwise they would effectively totally negate usefulness of user presets

@mnalis I suspected that this would be the case :)

However, for the reasons you put into a great concrete list in https://github.com/streetcomplete/StreetComplete/discussions/3209#discussioncomment-1225267, I don't think that *fully* customizable quest groups are workable. At least, not alongside quest presets. In particular, I think that the ability to create **custom groups** overlaps too closely with quest presets. Instead, I think a fixed set of **pre-defined** groups, possibly allowing the use to customize which quests are in them, might be useful.

#### Subquests

These would be pre-defined groups with non-customizable membership (maybe allowing enable/disable individually). The goal is to reduce the length of the list. Basically, for situations where they are *technically* implemented as different quests, but are really the same.
I think there are **not many** useful groups of this type, because it's too hard to create one which nobody will want to split apart.

- Maybe things like the many variations of "Is this ____ lit?" Or, I think there was talk of splitting the [smoothness quest](https://github.com/streetcomplete/StreetComplete/pull/3257) by way type.
- Alternatively, groups of quests that you'd almost always solve in sequence, like the diet quests (it was even discussed to combine them into one form). The parking quests might also fit here.
- Edit: Something like https://github.com/streetcomplete/StreetComplete/issues/3291 is probably the best example. The ambiguous tagging scenario could be split into two subquests but still shown as one in the UI.

I'm not sure how useful these would be, so I'm **not going to propose** it at this time.

#### Favorites

These would be pre-defined groups with customizable membership. The idea is that instead of comparing each quest against each other quest, you just need to assign it to a priority bucket. It's a simpler (I think) version of https://github.com/streetcomplete/StreetComplete/issues/3034#issuecomment-879481130.

- Remove the ability to put quests in an arbitrary order.
- Add the ability to mark quests as favorites. Favorites are always sorted at the top (and secondarily by the default order).
    - The app wouldn't describe it this way, but in effect this is "adding a quest to the *favorites* group".
    - If favorites, regular, and disabled are not fine-grained enough, there could be another "low priority" group.
- Favorites differ between presets. So for your biking preset, you'd favorite [these quests](https://github.com/streetcomplete/StreetComplete/discussions/2457#discussioncomment-268321).

I think I will end up proposing this in an issue. But I'm curious first about your opinion, as the relentless advocate for customizability :D. It reduces the granularity of your control over the quest order, but I wonder how much granularity is needed. From a different perspective: how many "buckets" would you need to accomplish your desired order? If you want, we could actually test this by implementing it with a number input instead, where higher numbers are sorted first and equal numbers are sorted in the default order.

**Reply by smichel17** [reply]
Personally I *think* I would want just 2 or 3 buckets (plus disabled):

1. Favorites, for quests like Address that I want to prioritize.
2. Regular, for most quests.
3. Low priority, for low impact ("spammy") quests I only want to see after I've completed everything else in the area (roof shape).
    - With quest presets, I think I don't need this, since I could make two presets, with the same favorites, where one has low priority quests disabled and the other has them enabled.
4. Disabled, for quests that I don't feel equipped to answer.

I'm not sure if each quest should have a 3-way menu between Favorites, Regular, and Disabled, or whether it should be two separate toggles to enable/disable and star/unstar. I'd have to play around with a prototype to get a feel for it, I think.

**Reply by mnalis** [reply]
> Remove the ability to put quests in an arbitrary order.

Needless to say, I'm horrified with that prospect :smile:  e.g. https://github.com/streetcomplete/StreetComplete/issues/3034#issuecomment-879995215

> Personally I think I would want just 2 or 3 buckets (plus disabled):

Those "buckets" in that case could be accomplished with just regular `presets`, which we already have (they'd just need _one-time effort_ to initialize them to user liking). And all that remains then is a `quick preset switch` button on main screen, e.g. https://github.com/streetcomplete/StreetComplete/issues/3827 which is sorely missed (along with an option to import/export quest settings and share them between users https://github.com/streetcomplete/StreetComplete/issues/3481, but I digress)

@Helium314 fork https://github.com/Helium314/StreetComplete implements similar button, where one can change current `preset` in just `3` clicks (instead of quite cumbersome `9` clicks in upstream SC). 
If the button in upstream SC were used only for that functionality, it could be reduced to just `2` clicks which would be great - and as mentioned in that issue the button needs not be shown if user has just default preset, so the screen estate is only used for users who really want it)

**Comment by timothywashere** [comment]
Personally I would like quest groups, I find the huge list of quests unmanageable.

I would offer the user quest groups by default but have an option to turn off groups if you want full control.

To me it would work like this. 
Each quest would have one group only. Bicycle parking would be in cycling because it's only really interesting to cyclists. I doubt there are many parking enthusiasts out there.

The group screen would offer a list that can be reorderer and would have a high and low priority for each group. You could reorder them all except you couldn't put a Low Priority above the High Priority of the same group.

High Priority Cycling
High Priority Buildings
Low Priority Cycling 
Low Priority Buildings
High Priority Post Box
Low Priority Post Box 

Clicking on either a High or Low priority group would launch a separate screen which would show all of the quests within that group and offer you three radio buttons for each and you select one of them.

Bicycle Parking   High(0)  Low(x)  Off(0)
Bicycle Lanes      High(x)  Low(0)  Off(0)

**Reply by mnalis** [reply]
sounds interesting, but when I try to look deeper into it, I hit various blocks:
- making groups/quests three-state instead of boolean makes them even more cumbersome
- having to configure each of the groups makes it more complicated too (unless one happens to be happy with defaults; but then, if we were kind of people who are happy with defaults we wouldn't be having this conversation :smile:) 
- looking at each quest and trying to categorize it, I think there would be  a **lots of groups** - and since there are two types of each (high and low priority) is it quite possible that at the end the number of groups wouldn't be all that much smaller than number of quests, *and* you'd have to additionally configure each of them! Unless the number of groups is limited to 5-6 groups (thus 10-12 group choices), it would likely turn to be more complex than current situation.

- some things seem to belong to more than one group. e.g. would `surface` / `surface quality` be in `cycling` or `car` group?  If you ask me, definitely in `cycling` - but that's just because I don't care about cars... how about `lit` quest? etc.
- since some quests might be logically in few different groups, it would also be much harder to find them
- also, depending on my free time, I can't seem to clear cut my answers in "I want to be solving high priority building quests" group. Sometimes I want to solve high-priority `address quests`, but that has a lots of elements, so I sometimes decide not solve those, but only solve little less important (but still important) but much rarer (when outside of big city) quest like "opening hours" (which is important outside of the city, as it is often only place when you can get drink/food in such areas). 

So, quests preset work for me _mostly_... What I'd definitively like is the ability so export/import them (also allowing to clone current preset and make changes to the clone); and to change them easier (current 9-clicks gesture gets annoying really fast). 

I'd also love easier UI to more easily find the quests, which grouping in a sense of "expand this category" might provide, but it is less important as I make modifications to my presets much more rarely. But they are quite a pain to set them up the first time, that is true.

**Reply by Helium314** [reply]
> also allowing to clone current preset and make changes to the clone

Did your try https://github.com/Helium314/StreetComplete/commit/1195579ab09d47f319dc883810c47a4240493d14?

**Reply by mnalis** [reply]
I have tried it now, works well! (I initially had an simpler idea, just always copy currently selected preset when creating new one, but this works nice also). Last time I tried new profile in your fork was in pre-official-preset-days when you did first `presets` implementation (which also had copy-from-profile feature!). I'm unfortunately currently lagging a lot regarding your quite interesting fork @Helium314; hopefully I'll give it some more deserved attention when I cleanup my own fork and propose/submit few things I have in mind in official SC.

**Reply by smichel17** [reply]
> I can't seem to clear cut my answers in "I want to be solving high priority building quests" group. Sometimes I want to solve high-priority `address quests`, but that has a lots of elements, so I sometimes decide not solve those, but only solve little less important (but still important) but much rarer (when outside of big city) quest like "opening hours" (which is important outside of the city, as it is often only place when you can get drink/food in such areas).

I wonder how much of this could be addressed by better controls around the density of displayed quests. E.g. https://github.com/streetcomplete/StreetComplete/issues/1654#issuecomment-741970370, https://github.com/streetcomplete/StreetComplete/discussions/2457#discussioncomment-267405

And/or, by a more easily accessible interface for changing visible quests. E.g. if there were a menu item, which opened a sidebar similar to the undo sidebar, where you could easily toggle groups of quests or their order. Or something like https://github.com/streetcomplete/StreetComplete/issues/2944, https://github.com/streetcomplete/StreetComplete/issues/3034#issuecomment-879481130=.  If you're only making minor tweaks, like "which of these few important quests should be shown first", then it's the same amount of effort to just make those tweaks every time; and it's probably a better fit for what you're trying to do, since you don't have to create a new preset each time you want some unique order.
